### PR TITLE
Added generateCommand.py, which generates pipeline commands from YAML.

### PR DESCRIPTION
**Pull request by SogoMineo**
The script generateCommand.py generates pipeline commands
that can be executed by bash:
    source <(generateCommand.py ...)
Inputs and configurations of the pipeline commands are described
in a YAML file.

### CODE REVIEW

File: bin.src/generateCommands.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+        print(f"{shellCommand(command)} || return 1")
+
+
+class CalibSource:
+    __slots__ = ["configs", "visits"]
+    __subclasses:Dict[str,type] = {}
+
+    def __init__(self, configs:Iterator[str], visits:Iterator[int]):
+        self.configs = list(configs)
+        self.visits  = list(visits)
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:153]: Why does it have to take a list of visits? We can be so much more expressive about how to select visits than just giving a number, e.g., `--id field=BIAS dateObs=2020-05-13`. I would hate to lose this capability!

File: bin.src/generateCommands.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+            "--id", f"visit={visits}",
+            "-j", processes,
+            ]
+
+        if self.configs:
+            command += ["--config"] + self.configs
+
+        print(f"{shellCommand(command)} || return 1")
+
+        #ingestCalibs(self.typeName, dataDir, calib, rerun, copyMode)
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:277]: Remove commented-out code.

File: bin.src/generateCommands.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+        visits = visitsToString(self.visits)
+
+        command = [
+            self.commandName,
+            dataDir,
+            f"--calib={os.path.join(dataDir, calib)}",
+            f"--rerun={rerun}",
+            "--id", f"visit={visits}",
+            "-j", processes,
+            ]
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:270]: This seems to be a common pattern. Factor it out?

File: bin.src/generateCommands.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+            command += ["--config"] + self.configs
+
+        print(f"{shellCommand(command)} || return 1")
+
+        #ingestCalibs(self.typeName, dataDir, calib, rerun, copyMode)
+        ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
+
+
+class CalibBlock:
+    calibTypes = [] # will be replaced in processYaml()
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:282]: I think this breaks encapsulation. An external function shouldn't be responsible for changing this class' innards.

File: bin.src/generateCommands.py [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+
+    def __init__(self, configs:Optional[Iterator[str]]):
+        self.configs = list(configs) if configs else []
+
+    @classmethod
+    def fromYaml(cls, yamlBlock:Mapping[str,Any]) -> "ScienceStep":
+        r"""
+        \param yamlBlock: {
+            "config": ...,
+            "configs": ...,
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:320]: This syntax has strong potential for confusion.

File: bin.src/generateCommands.py [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+            }
+        """
+        configs = _getConfigs(yamlBlock)
+        return cls(configs)
+
+    @staticmethod
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:326]: Surely this is a `classmethod`? But I'm hoping it will go away when you use `__subclasses__`.

File: bin.src/generateCommands.py [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+    def getSubclass(typeName:str) -> type:
+        return ScienceStep.__subclasses[typeName]
+
+    @staticmethod
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:330]: Ditto.

File: bin.src/generateCommands.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+    def registerAsSubclass(typeName:str):
+        def helper(cls):
+            ScienceStep.__subclasses[typeName] = cls
+            cls.typeName = typeName
+            return cls
+        return helper
+
+    # To use (subclass instance).execute(),
+    # remember to override `commandName`
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:339]: Comment needs to be in the class docstring.

File: bin.src/generateCommands.py [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+
+    if "configs" in yamlBlock:
+        for c in yamlBlock["configs"]:
+            configs.append(c)
+
+    return configs
+
+def _getVisits(yamlBlock:Mapping[str,Any]) -> List[int]:
+    """
+    Extract visit: ... and visits: ... from yamlBlock.
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:422]: This is also potentially confusing syntax.

File: bin.src/generateCommands.py [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+
+    return visits
+
+
+def ingestCalibs(calibType, dataDir, calib, rerun, copyMode, *, overwrite=False):
+    command = [
+        "ingestCalibs.py",
+        dataDir,
+        f"--output={os.path.join(dataDir, calib)}",
+        "--validity=1800",
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:456]: We must be able to specify the validity range, and it should be allowed to be different for different calib types.

File: policy/LAM/integration_test.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
+#
```
Comment (PaulPrice) [inline policy/LAM/integration_test.yaml:1]: Why is this in the `policy` directory?
What does this have to do with `LAM`?

File: policy/LAM/integration_test.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  dirName : "$DRP_STELLA_DATA_DIR/raw"
+  arms : [r1, b1]
+  detectorMapFmt : detectorMap-sim-%s.fits
+#
+# Blocks of data which are used together to update the calibrations
+#
+calibBlocks :
+  integration_test :
+    bias :
+      visits : ["0..4"]
```
Comment (PaulPrice) [inline policy/LAM/integration_test.yaml:31]: Why is this a list? Shouldn't it just be a string? Or do different elements in the list result in different executions of `constructBias`?

File: policy/LAM/integration_test.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  [
+  reduceExposure, mergeArms, calculateReferenceFlux, fluxCalibrate, coaddSpectra
+  ]
+#
+# Bootstrap the calibration directory
+#
+# The files are to be found in dirName, taken to be relative to --dataDir if
+# not an absolute path
+#
+bootstrap :
```
Comment (PaulPrice) [inline policy/LAM/integration_test.yaml:21]: No space before the colon.

File: bin.src/generateCommands.py [hunk 14]
```diff
@@ -0,0 +1,581 @@
...
+    Process a yaml file and generate the desired calibrations
+    """)
+
+    parser.add_argument('dataDir', type=str, help="Root of data repository")
+    parser.add_argument('specFile', type=str, help="Name of file specifying the work")
+    parser.add_argument('--bootstrap', action="store_true", help="Bootstrap the calibrations?")
+    parser.add_argument('--blocks', type=str, nargs="+", help="Blocks to execute")
+    parser.add_argument('--calib', type=str, help="Name of output calibration directory", default="CALIB")
+    parser.add_argument('--calibTypes', type=str, nargs="+", help="Types of calibs to process")
+    parser.add_argument("--copyMode", choices=["move", "copy", "link", "skip"], default="link",
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:46]: I don't think `link` is always the best. `copy` is probably the best default, but I'm concerned that the best choice may change as a function of calib type.

File: bin.src/generateCommands.py [hunk 15]
```diff
@@ -0,0 +1,581 @@
...
+                        help="How to move files into calibration directory")
+    parser.add_argument('--force', action="store_true", help="Continue in the face of problems", default=False)
+    parser.add_argument("-j", "--processes", type=int, default=1, help="Number of processes to use")
+    parser.add_argument('--rerun', type=str, help="Name of rerun to use for calib processing")
+    parser.add_argument('--scienceSteps', type=str, nargs="+", help="pipeline steps to execute")
+    parser.add_argument('--verbose', action="store_true", help="How chatty should I be?", default=False)
```
Comment (PaulPrice) [inline bin.src/generateCommands.py:52]: Add an option to clean up calib reruns (save disk space; this should probably be the default).
Add an option `--developer` to add `--no-versions --clobber-config` to commands.

File: policy/LAM/integration_test.yaml (hunk 11 above)
Comment (SogoMineo) [reply policy/LAM/integration_test.yaml:1]: There is no reason for this choice. Where should it be placed?

File: bin.src/generateCommands.py [hunk 16]
```diff
@@ -0,0 +1,581 @@
...
+            ]
+
+        command += detectorMaps
+
+        print(f"{shellCommand(command)} || return 1")
+
+
+class CalibSource:
+    __slots__ = ["configs", "visits"]
+    __subclasses:Dict[str,type] = {}
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:149]: I didn't know the existence of `__subclasses__`... It does exist, but I cannot find its documentation. Is it not CPython's documentation detail? But I think I should at least use `__init_subclass__`, which is documented.

File: bin.src/generateCommands.py (hunk 5 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py:320]: Only "config" key should exist, and its value should be of type `Union[int, str, List[int, str]]` . Is this correct?

File: bin.src/generateCommands.py [hunk 17]
```diff
@@ -0,0 +1,581 @@
...
+
+from typing import Any, Dict, Iterator, List, Mapping, Optional, Sequence, Tuple
+
+
+def main():
+    parser = argparse.ArgumentParser(description="""
+    Process a yaml file and generate the desired calibrations
+    """)
+
+    parser.add_argument('dataDir', type=str, help="Root of data repository")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:40]: I'm replacing `'` to `"`. I never type `'` for myself because `'` (SHIFT+7) is far less accessible than `"` (SHIFT+2) in a jp106 keyboard. Occurrences of `'` were almost surely copied from somewhere.

File: bin.src/generateCommands.py [hunk 18]
```diff
@@ -0,0 +1,581 @@
...
+    parser.add_argument('specFile', type=str, help="Name of file specifying the work")
+    parser.add_argument('--bootstrap', action="store_true", help="Bootstrap the calibrations?")
+    parser.add_argument('--blocks', type=str, nargs="+", help="Blocks to execute")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:43]: "Blocks" are named arbitrarily in an input YAML file. Choices are not known beforehand.

File: bin.src/generateCommands.py [hunk 19]
```diff
@@ -0,0 +1,581 @@
...
+    parser.add_argument('--calib', type=str, help="Name of output calibration directory", default="CALIB")
+    parser.add_argument('--calibTypes', type=str, nargs="+", help="Types of calibs to process")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:45]: I'm following the original design that the choices of calibTypes are described in a YAML file. But I agree the choices can be hardcoded. If so, they should be hardcoded.

File: policy/LAM/integration_test.yaml (hunk 11 above)
Comment (PaulPrice) [reply policy/LAM/integration_test.yaml:1]: I'm not sure. Maybe under `examples` or `etc`?

File: bin.src/generateCommands.py (hunk 5 above)
Comment (PaulPrice) [reply bin.src/generateCommands.py:320]: It could be a `float` too.
Could you add an option for providing a `--configfile`, which would give us full control?

Comment (SogoMineo) [reply bin.src/generateCommands.py:320]: > `Union[int, str, List[int, str]]`

I confused "config" with "visit". For "config" key, the type of its value should be `Union[str, List[str]]`.

> `--configfile`

If they were to use `--configfile`, the YAML file would no longer be defining a unique (input, config) set. Is it OK?

Comment (PaulPrice) [reply bin.src/generateCommands.py:320]: I'm not sure I see the problem with that.
I think we will want to use the scripts you generate to drive experiments that involve changing configuration options that can't be changed just with `--config`.

File: bin.src/generateCommands.py [hunk 20]
```diff
@@ -0,0 +1,581 @@
...
+    bootstrap, calibBlocks, scienceBlocks = processYaml(args.specFile)
+    possibleBlocks = unique(list(calibBlocks.keys()) + list(scienceBlocks.keys()))
+    if args.blocks is None:
+        args.blocks = possibleBlocks
+
+    args.dataDir = os.path.abspath(args.dataDir)
+    if not os.path.exists(args.dataDir):
+        print(f"Warning: {args.dataDir} doesn't exist", file=sys.stderr)
+        if not args.force:
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:65]: Though I don't exactly understand the purpose of '--force' (it existed in the original code), it is in fact useful here because users can generate bash commands they don't execute immediately, without `mkdir $dataDir`.

File: bin.src/generateCommands.py [hunk 21]
```diff
@@ -0,0 +1,581 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py:1]: Moved all codes to python/.

File: bin.src/generateCommands.py [hunk 22]
```diff
@@ -0,0 +1,581 @@
...
+import shlex
+import sys
+import textwrap
+
+import yaml
+
+from typing import Any, Dict, Iterator, List, Mapping, Optional, Sequence, Tuple
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Wrote docstrings in numpydoc format.

[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: In the new revision, main() does only argument parsing. The other things has gone to generateCommands().

File: bin.src/generateCommands.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Added choices.

File: bin.src/generateCommands.py (hunk 14 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: I changed the default option to `copy`. This option has ceased to influence the copy mode of the initial detectorMap. It is now always `copy` no matter what this option is.

File: bin.src/generateCommands.py [hunk 23]
```diff
@@ -0,0 +1,581 @@
...
+    parser.add_argument('--bootstrap', action="store_true", help="Bootstrap the calibrations?")
+    parser.add_argument('--blocks', type=str, nargs="+", help="Blocks to execute")
+    parser.add_argument('--calib', type=str, help="Name of output calibration directory", default="CALIB")
+    parser.add_argument('--calibTypes', type=str, nargs="+", help="Types of calibs to process")
+    parser.add_argument("--copyMode", choices=["move", "copy", "link", "skip"], default="link",
+                        help="How to move files into calibration directory")
+    parser.add_argument('--force', action="store_true", help="Continue in the face of problems", default=False)
+    parser.add_argument("-j", "--processes", type=int, default=1, help="Number of processes to use")
+    parser.add_argument('--rerun', type=str, help="Name of rerun to use for calib processing")
+    parser.add_argument('--scienceSteps', type=str, nargs="+", help="pipeline steps to execute")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Added choices.

File: bin.src/generateCommands.py (hunk 15 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Added option `--clean`.

File: bin.src/generateCommands.py [hunk 24]
```diff
@@ -0,0 +1,581 @@
...
+    args = parser.parse_args()
+
+    bootstrap, calibBlocks, scienceBlocks = processYaml(args.specFile)
+    possibleBlocks = unique(list(calibBlocks.keys()) + list(scienceBlocks.keys()))
+    if args.blocks is None:
+        args.blocks = possibleBlocks
+
+    args.dataDir = os.path.abspath(args.dataDir)
+    if not os.path.exists(args.dataDir):
+        print(f"Warning: {args.dataDir} doesn't exist", file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Replaced `print(..., file=sys.stderr)` to `lsst.log`

File: bin.src/generateCommands.py [hunk 25]
```diff
@@ -0,0 +1,581 @@
...
+            sys.exit(1)
+
+    unrecognisedBlocks = set(args.blocks) - set(possibleBlocks)
+    if unrecognisedBlocks:
+        print(f"Unrecognised blocks: '{list(unrecognisedBlocks)}'%s" % (" ignoring" if args.force else ""),
+              file=sys.stderr)
+        if args.verbose:
+            print(f"Possible blocks are {list(possibleBlocks)}", file=sys.stderr)
+        if not args.force:
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Now that `generateCommands()` function has been split from `main()` , this block may be called from other functions than `main()`. Because the caller may pass invalid arguments to `generateCommands()`, this block still remains.

File: bin.src/generateCommands.py [hunk 26]
```diff
@@ -0,0 +1,581 @@
...
+
+    #
+    # Bootstrap the calibrations
+    #
+    if args.bootstrap:
+        bootstrap.execute(
+            args.dataDir, args.calib, args.copyMode,
+            verbose=args.verbose
+            )
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Renamed "bootstrap" to "init".

File: bin.src/generateCommands.py [hunk 27]
```diff
@@ -0,0 +1,581 @@
...
+
+    #
+    # Process the main blocks
+    #
+
+    for blockName in args.blocks:
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: "Blocks" has names named by YAML writers. The order of blocks in the command line (not YAML) is respected because users may expect blocks to be processed in the order they write.

File: bin.src/generateCommands.py [hunk 28]
```diff
@@ -0,0 +1,581 @@
...
+            scienceBlocks[blockName].execute(
+                args.dataDir, args.calib, args.rerun,
+                steps=args.scienceSteps, processes=args.processes, verbose=args.verbose
+                )
+
+
+class BootstrapSource:
+    __slots__ = ["dirName", "detectorMapFmt", "arms"]
+
+    def __init__(self, dirName:str, detectorMapFmt:str, arms:Iterator[str]):
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Pacified flake8.

[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: (But I believe `def f(param1: Dict[str, str] = {}, para2: Dict[str, str] = {})` (flake8-clean) is no more readable than `def f(param1 = {}, param2 = {})` (flake8-unclean) is)

File: bin.src/generateCommands.py [hunk 29]
```diff
@@ -0,0 +1,581 @@
...
+        self.dirName        = str(dirName)
+        self.detectorMapFmt = str(detectorMapFmt)
+        self.arms           = list(arms)
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: `arms` has to be copied because its parameter type is `Iterator` . Hence `list(arms)` . The others were cast for an aesthetic reason. Only `arms` is cast now.

File: bin.src/generateCommands.py [hunk 30]
```diff
@@ -0,0 +1,581 @@
...
+        bootstrapDir = os.path.join(dataDir, os.path.expandvars(self.dirName))
+        if verbose:
+            print("Reading bootstrap files from %s" % (bootstrapDir), file=sys.stderr)
+
+        detectorMaps = []
+        for arm in self.arms:
+            detectorMaps.append(os.path.join(bootstrapDir, self.detectorMapFmt % arm))
+
+        command = [
+            "ingestCalibs.py",
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Renamed "bootstrap" to "init". I have not yet thought about `bootstrapDetectorMap.py` very well.

File: bin.src/generateCommands.py [hunk 31]
```diff
@@ -0,0 +1,581 @@
...
+            f"--output={os.path.join(dataDir, calib)}",
+            "--validity=1800",
+            "--create",
+            f"--mode={copyMode}",
+            "--",
+            ]
+
+        command += detectorMaps
+
+        print(f"{shellCommand(command)} || return 1")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Changed the default output to "a.sh". (This name can be changed with "--output" option)

File: bin.src/generateCommands.py (hunk 1 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Made `field` available. It is easy to add other keys like `dateObs`, but I have not done it yet.

File: bin.src/generateCommands.py [hunk 32]
```diff
@@ -0,0 +1,581 @@
...
+    @staticmethod
+    def registerAsSubclass(typeName:str):
+        def helper(cls):
+            CalibSource.__subclasses[typeName] = cls
+            cls.typeName = typeName
+            return cls
+        return helper
+
+    # To use (subclass instance).execute(),
+    # remember to override `commandName`
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Removed this comment. Now it is simply impossible to inherit this class without specifying `commandName`.

File: bin.src/generateCommands.py [hunk 33]
```diff
@@ -0,0 +1,581 @@
...
+    def execute(self, dataDir, calib, rerun, copyMode, processes=1, verbose=False):
+        visits = visitsToString(self.visits)
+
+        command = [
+            self.commandName,
+            dataDir,
+            f"--calib={os.path.join(dataDir, calib)}",
+            f"--rerun={rerun}",
+            "--id", f"visit={visits}",
+            "--batch-type=none", "-j", processes,
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Changed to --batch-type=smp.

File: bin.src/generateCommands.py [hunk 34]
```diff
@@ -0,0 +1,581 @@
...
+@CalibSource.registerAsSubclass("flat")
+class FlatSource(CalibSource):
+    commandName = "constructFiberFlat.py"
+
+@CalibSource.registerAsSubclass("fiberTrace")
+class FiberTraceSource(CalibSource):
+    commandName = "constructFiberTrace.py"
+
+    def execute(self, dataDir, calib, rerun, copyMode, processes=1, verbose=False):
+        for v in self.visits:
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Changed how to run `constructFiberTrace.py`. Accordingly, `fiberTrace` field in the YAML file now has a structure different from other calib types.

File: bin.src/generateCommands.py [hunk 35]
```diff
@@ -0,0 +1,581 @@
...
+                "--id", f"visit={v}",
+                "--batch-type=none", "-j", processes,
+                ]
+
+            if self.configs:
+                command += ["--config"] + self.configs
+
+            print(f"{shellCommand(command)} || return 1")
+
+        fiberTraceDir = os.path.join(dataDir, "rerun", str(rerun), "FIBERTRACE")
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: It was possible that rerun=None. Now, the type of "rerun" is specified as `str`, and programmers is to blame if rerun=None here.

File: bin.src/generateCommands.py [hunk 36]
```diff
@@ -0,0 +1,581 @@
...
+
+        for detector in ["b1", "r1"]:
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Added `m` and `n`.

File: bin.src/generateCommands.py [hunk 37]
```diff
@@ -0,0 +1,581 @@
...
+            command = textwrap.dedent(fr"""
+                trace0="$(find {shlex.quote(fiberTraceDir)} -maxdepth 1 -name 'pfsFiberTrace-*-{detector}.fits' -print -quit)"
+                if [ -n "$trace0" ]
+                then
+                    mkdir -p {shlex.quote(fiberTraceDir)}/COMBINED
+                    combineFiberTraces.py \
+                        {shlex.quote(fiberTraceDir)}/COMBINED/"$(basename "$trace0")" \
+                        $(find {shlex.quote(fiberTraceDir)} -maxdepth 1 -name 'pfsFiberTrace-*-{detector}.fits' -print) || return 1
+                fi
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Moved `mkdir` out of the "if" block. In addition, changed how to glob files so that non-posix `find` is no longer required.

File: bin.src/generateCommands.py [hunk 38]
```diff
@@ -0,0 +1,581 @@
...
+            """)
+            print(command[1:].rstrip())
+
+        ingestCalibs("FIBERTRACE/COMBINED", dataDir, calib, rerun, copyMode)
+        #ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Removed.

File: bin.src/generateCommands.py (hunk 3 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: This one has "-j {processes}" in it. The others has "--batch-type=smp ..." in it.

File: bin.src/generateCommands.py (hunk 2 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Removed.

File: bin.src/generateCommands.py (hunk 4 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: I agree. The current code does not do this.

File: bin.src/generateCommands.py (hunk 5 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Added `configfile` key.

File: bin.src/generateCommands.py (hunk 6 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: This has to be `classmethod` because `Subclass.fromYaml()` has to return `Subclass` instead of `ScienceStep`. The `cls` argument is essential.

File: bin.src/generateCommands.py (hunk 9 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: `visit` and `visits` have been merged to `visit.

File: bin.src/generateCommands.py (hunk 10 above)
Comment (SogoMineo) [reply bin.src/generateCommands.py]: Now `validity` can be written in the YAML file. If not specified, it defaults to 1800.

File: bin.src/generateCommands.py [hunk 39]
```diff
@@ -0,0 +1,581 @@
...
+
+
+def shellCommand(command:Iterator[str]) -> str:
+    return ' '.join(shlex.quote(str(x)) for x in command)
+
+
+def unique(seq:Iterator[Any]) -> List[Any]:
+    """
+    Get unique elements in `seq` .
+    The order of the elements will be retained.
```
[the remark this answers is not in the document]
Comment (SogoMineo) [reply bin.src/generateCommands.py]: It is random. (But I don't think it is important because I have not included `unique()` in `__all__`, and discouraged other people from using it.)

File: policy/LAM/integration_test.yaml (hunk 11 above)
Comment (SogoMineo) [reply policy/LAM/integration_test.yaml]: This file has been moved to `examples/`.

File: policy/LAM/integration_test.yaml (hunk 12 above)
Comment (SogoMineo) [reply policy/LAM/integration_test.yaml]: Originally, this had to be a list because of `visit` and `visits` distinction. Now these two keys has been merged to `visit`, and this value does not have to be a list.

File: python/pfs/pipe2d/generateCommands.py [hunk 40]
```diff
@@ -0,0 +1,1493 @@
...
+    parser.add_argument("--copyMode", choices=["move", "copy", "link", "skip"], default="copy",
+                        help="How to move files into calibration directory")
+    parser.add_argument("--devel", action="store_true",
+                        help="Run commands in the development mode (no versioning)")
+    parser.add_argument("--force", action="store_true",
+                        help="Continue in the face of problems")
+    parser.add_argument("-j", "--processes", type=int, default=1, help="Number of processes to use")
+    parser.add_argument("-L", "--loglevel", type=str, choices=list(LogLevel.__members__), default="INFO",
+                        help="How chatty should I be?")
+    parser.add_argument("-o", "--output", type=str, default="a.sh", help="Path to output file")
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:80]: Can we just make this a required argument, instead of an obscure (but not without precedent, I know) default?

File: python/pfs/pipe2d/generateCommands.py [hunk 41]
```diff
@@ -0,0 +1,1493 @@
...
+        -------
+        options : `List[str]`
+            Like ["--config", "key=value", "key=value"]
+        """
+        commands = []
+        if self.configfile:
+            commands.append(f"--configfile={self.configfile}")
+        if self.configs:
+            commands += ["--config"] + self.configs
+        return commands
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:170]: Nicely done!

File: python/pfs/pipe2d/generateCommands.py [hunk 42]
```diff
@@ -0,0 +1,1493 @@
...
+        """
+        id = yamlBlock.get(key, [])
+        if isinstance(id, str):
+            id = [id]
+        if not isinstance(id, list):
+            raise RuntimeError(f"'{key}' must be a string or a list of string: '{id}'")
+
+        return cls(ensureKeyEqValue(str(x)) for x in id)
+
+    def getCommandLine(self, *, key: str = "id") -> List[str]:
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:220]: Pretty sure spaces around the equals sign are forbidden in the style guide, but not sure how that gets altered when there's types specified. Please make sure that this file passes flake8 with the usual configuration (which you may have to look in drp_stella for).

File: python/pfs/pipe2d/generateCommands.py [hunk 43]
```diff
@@ -0,0 +1,1493 @@
...
+            Logger.
+        fout : `TextIO`
+            Output file where to write the command.
+        dataDir : `str`
+            Root of data repository.
+        calib : `str`
+            Name of output calibration directory.
+        """
+        initDir = os.path.join(dataDir, os.path.expandvars(self.dirName))
+        logger.info(f"Reading init files from {initDir}")
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:301]: When using the LSST logger, you should use the old `%` formatting, and include the variables as parameters, rather than interpolating them. This allows the code to only do the string manipulation if it's warranted by the log level. (Not too important here, but a good habit to get into.)

    logger.info("Reading init files from %s", initDir)

File: python/pfs/pipe2d/generateCommands.py [hunk 44]
```diff
@@ -0,0 +1,1493 @@
...
+
+        Returns
+        -------
+        subclass : type
+            The class object of the subclass.
+        """
+        return CalibSource.__subclasses[typeName]
+
+    def execute(
+            self, logger: lsst.log.Log, fout: TextIO, dataDir: str, calib: str, rerun: str,
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:412]: All the methods take a `logger`. Maybe provide once to the constructor instead?

File: python/pfs/pipe2d/generateCommands.py [hunk 45]
```diff
@@ -0,0 +1,1493 @@
...
+        yamlBlock : `Mapping[str, Any]`
+            A block of a YAML data structure:
+
+            ``"group"``
+                List of Mapping[str, Any]`.
+                fiber traces are created for each group, and then combined.
+                Contents of each group are:
+
+                ``"id"``
+                    A string like ``"field=BIAS"`` or a list of such strings.
```
Comment (PaulPrice) [inline python/pfs/pipe2d/generateCommands.py:715]: Can I have something like:

    id:
        - "field=FLAT_ODD dateObs=2020-05-29"
        - "field=FLAT_EVEN dateObs=2020-05-29"

?

File: python/pfs/pipe2d/generateCommands.py (hunk 42 above)
Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py:220]: I don't like this style either, but flake8 requires spaces around `=` after a type annotation. If I defy this rule, flake8 raises E252 "missing whitespace around parameter equals." (See https://github.com/PyCQA/pycodestyle/issues/760) It is true this error code is not listed here: https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes but the latest code https://github.com/PyCQA/pycodestyle/blob/master/pycodestyle.py#L1000 actually checks it.

File: python/pfs/pipe2d/generateCommands.py (hunk 45 above)
Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py:715]: If such is the syntax, and if one writes `id: ["field=FLAT_ODD", "dateObs=2020-05-29"]`, then the translation result will be `--id field=FLAT_ODD` for one group and `--id dateObs=2020-05-29` for another. This is inconsistent with other calibTypes. If I, for consistency, forbid `List[str]` from being the type of `id` field of other calibTypes, `id` field will be inconsistent to `config` field. I don't want `config` field to be a single string like `config: "key=value key=value"` because such a string has to be parsed and split by spaces when `value` may contain spaces in theory.

File: python/pfs/pipe2d/generateCommands.py (hunk 44 above)
Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py:412]: Does it matter if I remove all `logger` parameters from `CalibSource` and `ScienceStep`? They are not used now. It may still be worth for `fromYaml()` method to have `logger` parameter.

File: python/pfs/pipe2d/generateCommands.py (hunk 40 above)
Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py]: I have promoted `--output` to the third positional parameter. I could keep it an option parameter, just making it required, but the document of `argparse` says it is bad mannered.

File: python/pfs/pipe2d/generateCommands.py (hunk 43 above)
Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py]: I have modified `logger` usages.

File: python/pfs/pipe2d/generateCommands.py (hunk 44 above)
Comment (PaulPrice) [reply python/pfs/pipe2d/generateCommands.py:412]: I don't have a strong opinion.

File: python/pfs/pipe2d/generateCommands.py (hunk 45 above)
Comment (PaulPrice) [reply python/pfs/pipe2d/generateCommands.py:715]: We need to be able to select data based on multiple fields in the registry, like `--id field=FLAT_ODD dateObs=2020-05-29".

Comment (SogoMineo) [reply python/pfs/pipe2d/generateCommands.py:715]: The current design has this syntax:
```
group:
    -
        id: ["field=FLAT_ODD", "dateObs=2020-05-29"]
    -
        id: ["field=FLAT_EVEN", "dateObs=2020-05-29"]
```
Maybe I should revise the docstring.

**Comment by SogoMineo** [comment]
I explained "id" field more verbosely in docstrings. I dropped `logger` parameter from some function parameters because it is not used currently, and I believe there will never be anything to report in these functions except for raising errors.
